### PR TITLE
Convolution, streamlined code for unoptimized case

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
Convolution was written using the old style. I rewrote the code using the current best practice. I also added testing of strides and dilation in the numerical tests. 

Still to do is the runtime handling of the padding flags. Will come in a subsequent PR.

I should also add that there were "confusion" between which constructor to use for the specific IndexExpr subclasses. In nearly all interesting cases, the specific subclass constructor where not used as the superclass one was used, thus not involving the "conversion" defined in the subclass constructor. I had to add the cross-product of all the sub-classes to make the code work. This is why you have now all of them, for example.

```C++
class LiteralIndexExpr : public IndexExpr {
//...
  LiteralIndexExpr(IndexExpr const &o);
  LiteralIndexExpr(UndefinedIndexExpr const &o);
  LiteralIndexExpr(LiteralIndexExpr const &o);
  LiteralIndexExpr(NonAffineIndexExpr const &o);
  LiteralIndexExpr(QuestionmarkIndexExpr const &o);
  LiteralIndexExpr(PredicateIndexExpr const &o);
  LiteralIndexExpr(AffineIndexExpr const &o);
  LiteralIndexExpr(DimIndexExpr const &o);
  LiteralIndexExpr(SymbolIndexExpr const &o);
//...
};
```

the result is that one can now easily "reactualizes" an index expression defined in an outer context in a new inner context, possibly using a different type of sub index expression. For example, in the new Conv code
```C++
// original definition
          IndexExpr co = g * SymbolIndexExpr(COPerGroup) + coPerGroup;

// use in an inner context 
                      SmallVector<IndexExpr, 4> filterAccessFct;
                      filterAccessFct.emplace_back(DimIndexExpr(co));
```

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
For ref, I used this code to verify the reference algo
```python
import numpy as np
import math
import torch

N=1
G=1
CI=2
HI=5
WI=6
CO=2
KH=2
KW=2
X = np.random.rand(N, CI, HI, WI)
W = np.random.rand(CO, int(CI/G), KH, KW)

XX = torch.from_numpy(X)
WW = torch.from_numpy(W)
nn = torch.nn.Conv2d(CI, CO, KH, padding=0, stride=0, dilation=2, groups=G, bias=False)
nn.weight.data = WW
YY = nn(XX)
Yref = YY.detach().numpy()
print(np.shape(Yref))

def my_floor(a, b):
    return int(math.floor((1.0*a)/(1.0*b)))
def my_ceil(a, b):
    return int(math.ceil((1.0*a)/(1.0*b)))

# ONNXOpHelper.cpp:getConvDimMap 
# agrees with https://oneapi-src.github.io/oneDNN/dev_guide_convolution.html
#    where their d is one less than ours (ours: default 1; theirs, default 0)
# agrees with https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html
def my_cond2D_spacial_size(I, p1, p2, K, d, s):
    p = p1+p2
    res = my_floor(I+p-(K-1)*d -1, s) + 1
    return int(res)

def my_conv2D_output_shape(X, W, pad, stride, dilation):
    (N, CI, HI, WI) = np.shape(X)
    (CO, CIperG, KH, KW) = np.shape(W)
    HO = my_cond2D_spacial_size(HI, pad[0], pad[2], KH, dilation[0], stride[0])
    WO = my_cond2D_spacial_size(WI, pad[1], pad[3], KW, dilation[1], stride[1])
    return (N, CO, HO, WO)

def val4(array, i0, i1, i2, i3):
    shape = np.shape(array)
    assert i0>=0 and i0<shape[0], "assert failed in first dim: 0<="+str(i0)+"<"+str(shape[0])
    assert i1>=0 and i1<shape[1], "assert failed in second dim: 0<="+str(i1)+"<"+str(shape[1])
    assert i2>=0 and i2<shape[2], "assert failed in third dim: 0<="+str(i2)+"<"+str(shape[2])
    assert i3>=0 and i3<shape[3], "assert failed in fourth dim: 0<="+str(i3)+"<"+str(shape[3])
    return array[i0, i1, i2, i3]
    
def my_conv2D(X, W, pad, stride, dilation, groups):
    output_shape = my_conv2D_output_shape(X, W, pad, stride, dilation)
    res = np.zeros(output_shape)
    (N, CI, HI, WI) = np.shape(X)
    (CO, CIPerG, KH, KW) = np.shape(W)
    (N, CO, HO, WO) = np.shape(res)
    (G, COPerG) = (groups, int(CO/groups))
    (ph, pw) = pad[0:2]
    (sh, sw) = stride
    (dh, dw) = dilation
    for n in range(N):
        for g in range(G):
            for coPerG in range(COPerG):
                co = g * COPerG + coPerG
                for ho in range(HO):
                    for wo in range(WO):
                        r = 0.0 # for res[n, co, ho, wo]
                        for ciPerG in range(CIPerG):
                            ci = g * CIPerG + ciPerG
                            for kh in range(KH):
                                for kw in range(KW):
                                    if (ho * sh + kh * dh - ph >= 0 and ho * sh + kh * dh - ph < HI and \
                                        wo * sw + kw * dw - pw >= 0 and wo * sw + kw * dw - pw < WI):
                                        r = r + val4(X, n, ci, ho * sh + kh * dh - ph, wo * sw + kw * dw - pw) * \
                                            val4(W, co, ciPerG, kh, kw)
                        res[n, co, ho, wo] = r
    return res

Yalex = my_conv2D(X, W, [2,2,2,2], [2,2], [2,2], 2)
print(np.shape(Yalex))
res = np.allclose(Yref, Yalex)
print("both computations are the same?", res)

def my_k_range(o, I, K, p, s, d):
    # translate ho * sh + kh * dh - ph>=0  and ho * sh + kh * dh - ph<HI into min/max for kh
    min_in_range = my_ceil(p - o * s, d)
    max_in_range = my_ceil(I + p - o * s, d)
    return (max(min_in_range, 0), min(max_in_range, K))

def my_conv2D_v2(X, W, pad, stride, dilation, groups):
    output_shape = my_conv2D_output_shape(X, W, pad, stride, dilation)
    res = np.zeros(output_shape)
    (N, CI, HI, WI) = np.shape(X)
    (CO, CIPerG, KH, KW) = np.shape(W)
    (N, CO, HO, WO) = np.shape(res)
    (G, COPerG) = (groups, int(CO/groups))
    (ph, pw) = pad[0:2]
    (sh, sw) = stride
    (dh, dw) = dilation
    for n in range(N):
        for g in range(G):
            for coPerG in range(COPerG):
                co = g * COPerG + coPerG
                for ho in range(HO):
                    for wo in range(WO):
                        r = 0.0 # for res[n, co, ho, wo]
                        (min_kh, max_kh) = my_k_range(ho, HI, KH, ph, sh, dh)
                        (min_kw, max_kw) = my_k_range(wo, WI, KW, pw, sw, dw)
                        for ciPerG in range(CIPerG):
                            ci = g * CIPerG + ciPerG
                            for kh in range(min_kh, max_kh):
                                for kw in range(min_kw, max_kw):
                                    r = r + val4(X, n, ci, ho * sh + kh * dh - ph, wo * sw + kw * dw - pw) * \
                                        val4(W, co, ciPerG, kh, kw)
                        res[n, co, ho, wo] = r
    return res

Yalex2 = my_conv2D_v2(X, W, [2,2,2,2], [2,2], [2,2], 2)
print(np.shape(Yalex2))
res2 = np.allclose(Yref, Yalex2)
print("both computations are the same?", res2)
```

File: src/Conversion/ONNXToKrnl/NN/Conv.cpp [hunk 1]
```diff
@@ -16,12 +15,228 @@
...
+                      DimIndexExpr n(outerIndices[0]);
+                      inputAccessFct.emplace_back(n);
+                      // ci = g * CIPerG + ciPerG
+                      DimIndexExpr ciPerG(redIndices[0]);
+                      IndexExpr ci =
+                          DimIndexExpr(g) * SymbolIndexExpr(CIPerGroup) +
+                          ciPerG;
+                      inputAccessFct.emplace_back(ci);
+                      for (int i = 0; i < spacialRank; ++i) {
+                        // for each spacial dims: access is o * s + k * d - p.
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/NN/Conv.cpp:191]: Is it better to move the computation for `o * s - p` outside the current loop? Perhaps, a similar movement for `ci = g * CIPerG + ciPerG`

Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToKrnl/NN/Conv.cpp:192]: Updated the code as requested. Not sure it is not caught by later optimizations in the backend, but it probably does not hurt to do it here.

Comment (tungld) [reply src/Conversion/ONNXToKrnl/NN/Conv.cpp:192]: Thank you for the update!

**Comment by tungld** [comment]
@jenkins-droid test this please

**Comment by tungld** [comment]
@gongsu832 Jenkins is not responding. Could you help take a look at it?

**Review by tungld — APPROVED** [review]
LGTM.

Thank you for modernizing the code!

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please
